source/common/upstream/original_dst_cluster.cc

```diff
@@ -166,48 +167,82 @@ OriginalDstCluster::OriginalDstCluster(
 }
 
 void OriginalDstCluster::addHost(HostSharedPtr& host) {
-  HostMapSharedPtr new_host_map = std::make_shared<HostMap>(*getCurrentHostMap());
-  auto pair = new_host_map->emplace(host->address()->asString(), host);
-  bool added = pair.second;
-  if (added) {
-    ENVOY_LOG(debug, "addHost() adding {}", host->address()->asString());
-    setHostMap(new_host_map);
-    // Given the current config, only EDS clusters support multiple priorities.
-    ASSERT(priority_set_.hostSetsPerPriority().size() == 1);
-    const auto& first_host_set = priority_set_.getOrCreateHostSet(0);
-    HostVectorSharedPtr all_hosts(new HostVector(first_host_set.hosts()));
-    all_hosts->emplace_back(host);
-    priority_set_.updateHosts(0,
-                              HostSetImpl::partitionHosts(all_hosts, HostsPerLocalityImpl::empty()),
-                              {}, {std::move(host)}, {}, absl::nullopt);
+  std::string address = host->address()->asString();
+  HostMultiMapSharedPtr new_host_map = std::make_shared<HostMultiMap>(*getCurrentHostMap());
+  auto it = new_host_map->find(address);
+  if (it != new_host_map->end()) {
+    it->second->hosts_.push_back(host);
+  } else {
+    new_host_map->emplace(address, std::make_shared<HostsForAddress>(host));
   }
```

Comment:
Sorry to be really annoying, but this code is very confusing and I just want to make sure the next person doesn't have to take as long to understand it. Can you add comments here also about these 2 cases in terms of which concurrency leads to what branch? I get it now but would rather it be more clear to the next reader.

Reply:
Ack. LMK if this is sufficient, I spent too much time grokking this logic myself.